ARCs/arc-0060.md

```diff
@@ -0,0 +1,233 @@
+---
+arc: 60
+title: Algorand Wallet Arbitrary Signing API
+description: API function for signing data
+author: Bruno Martins (@ehanoc)
+status: Draft
+type: Standards Track
+category: Interface
+created: 2024-10-27
+requires: 1
+---
+
+## Abstract
+
+This ARC proposes a standard for arbitrary data signing. It is designed to be a simple and flexible standard that can be used in a wide variety of applications.
+
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+> Comments like this are non-normative
+
+## Rationale
+
+Signing data is a common and critical operation. Users may need to sign data for multiple reasons (e.g. delegate signatures, DIDs, signing documents, authentication).
+
+Algorand wallets need a standard approach to byte signing to unlock self-custodial services and protect users from malicious and attack-prone signing workflows.
+
+This ARC provides a standard API for bytes signing. The API encodes byte arrays to be signed into well-structured JSON schemas together with additional metadata. It requires wallets to validate the signing inputs, notify users about what they are signing and warn them in case of dangerous signing requests.
+
+### Overview
+
+This ARC defines a function `signData(signingData, metadata)` for signing data.
+
+`signingData` is a `StdSigData` object composed of the signing `data` that instantiates a known JSON Schema and the `signer`'s public key.
+
+
+### Signing Flow
+
+When connected to a specific `domain` (i.e app or other identifier), the wallet will receive a request to sign some `data` along side some `authenticatedData`, which will look like some random bytes. With this information, the wallet should follow the following steps:
+
+1. Hash the `data` field with `sha256`.
+2. Knowing to what `domain` we are connected to, hash such value with `sha256` and compare it with the first 32 bytes of `authenticatedData`.
+    2.1. If the hashes do not match, the wallet **MUST** return an error.
+3. Append the `authenticatedData` to the resulting hash of the `data` field.
+4. Sign the result
+
+
+### `Scopes`
+
+Supported scopes are:
+
+- `AUTH` (1): This scope is used for authentication purposes. It is used to sign data that will be used to authenticate the user. The `data` field **MUST** be a JSON object with the following fields:
+    - `type`:  the type of operation (e.g., "SIWA", “webauthn.create”, “webauthn.get”, etc)
+    - `origin`: The origin / domain of the request or the URL of the relying party (RP) that initiated the authentication.
+    - `challenge`: 32 bytes of entropy.
+
+Summarized signing process for `AUTH` scope:
+```plaintext
+EdDSA(SHA256(data) + SHA256(authenticatedData))
```

Comment:
"prefix" are a really bad practice in my opinion; prevents any subject to validate signature crafted in algorand wallets / dapps and generically verify them without conforming to our own custom way of doing things. 

We can add a check to prevent "Program"; sure. But the chances of you finding a hash with valid 7 bytes (Program) + the remainder of the program are extremely low. 

Still, i can add a check to reject "Program"